internal/controller/apiextensions/composite/composed.go

```diff
@@ -238,3 +278,324 @@ func IsReady(_ context.Context, cd resource.Composed, t v1.ComposedTemplate) (bo
+// named composer, but using the anonymous composer may be surprising. There's a
+// risk that someone added a new anonymous template to a Composition that
+// otherwise uses named templates. If they added the new template to the
+// beginning or middle of the resources array using the anonymous composer would
+// be destructive, because it assumes template N always corresponds to existing
+// template N.
+func RejectMixedTemplates(comp *v1.Composition) error {
+	named := 0
+	for _, tmpl := range comp.Spec.Resources {
+		if tmpl.TemplateName != nil {
+			named++
+		}
+	}
+
+	// We're using only anonymous templates.
+	if named == 0 {
+		return nil
+	}
+
+	// We're using only named templates.
+	if named == len(comp.Spec.Resources) {
+		return nil
+	}
+
+	return errors.New(errMixed)
+}
+
+// RejectDuplicateNames validates that all template names are unique within the
+// supplied Composition.
+func RejectDuplicateNames(comp *v1.Composition) error {
+	seen := map[string]bool{}
+	for _, tmpl := range comp.Spec.Resources {
+		if tmpl.TemplateName == nil {
+			continue
+		}
+		if seen[*tmpl.TemplateName] {
+			return errors.New(errDuplicate)
+		}
+		seen[*tmpl.TemplateName] = true
+	}
+	return nil
+}
+
+type composedResource struct {
+	Renderer
+	ConnectionDetailsFetcher
+	ReadinessChecker
+}
+
+// An AnonymousComposer composes resources using a Composition with anonymous
+// (i.e. unnamed) resource templates. It relates composed resources to their
+// templates using their index within the composite resource's resource ref
+// array and within the composition's resource template array. It assumes the
+// length and order of the Composition is immutable; it does not garbage collect
+// resources when a template is deleted, and may leak resources when the array
+// of templates is reordered.
+type AnonymousComposer struct {
+	client   resource.ClientApplicator
+	composed composedResource
+}
+
+// Compose resources into the supplied composite per the supplied composition.
+func (r *AnonymousComposer) Compose(ctx context.Context, cr resource.Composite, comp *v1.Composition) (CompositionResult, error) {
+	refs := make([]corev1.ObjectReference, len(comp.Spec.Resources))
+	copy(refs, cr.GetResourceReferences())
+
+	cds := make([]*composed.Unstructured, len(refs))
+	for i := range refs {
+		cd := composed.New(composed.FromReference(refs[i]))
+		if err := r.composed.Render(ctx, cr, cd, comp.Spec.Resources[i]); err != nil {
+			return CompositionResult{}, errors.Wrapf(err, errFmtRenderIdx, i)
+		}
+
+		cds[i] = cd
+		refs[i] = *meta.ReferenceTo(cd, cd.GetObjectKind().GroupVersionKind())
+	}
+
+	cr.SetResourceReferences(refs)
+	if err := r.client.Update(ctx, cr); err != nil {
+		return CompositionResult{}, errors.Wrap(err, errUpdate)
+	}
+
+	conn := managed.ConnectionDetails{}
+	ready := 0
+	for i, cd := range cds {
+		if err := r.client.Apply(ctx, cd, resource.MustBeControllableBy(cr.GetUID())); err != nil {
```

Comment:
This logic feels duplicated between the two composers, but unsure whether something like `cdStatus, err := r.composed.Reconcile(ctx, cd, cr, tpl)` reduces the complexity.

Reply:
I agree - this whole loop is quite similar between the two Composers. In much of it the only difference is that we lookup the template by index vs by map key. I suspect we can do better here.